test/jdk/java/nio/file/Files/CopyMoveVariations.java

```diff
@@ -0,0 +1,234 @@
+import static java.nio.file.LinkOption.*;
+import static java.nio.file.StandardCopyOption.*;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class CopyMoveVariations {
+    enum OpType {
+        COPY,
+        MOVE
+    }
+
+    enum PathType {
+        FILE,
+        DIR,
+        LINK
+    }
+
+    private static Stream<Arguments> params() {
+        List<Arguments> list = new ArrayList<Arguments>();
+
+        boolean[] falseAndTrue = new boolean[] {false, true};
+        for (PathType type : PathType.values()) {
+            String[] modes = new String[] {
+                "---------", "r--r--r--", "-w--w--w-", "rw-rw-rw-"
+            };
+            for (String mode : modes) {
+                for (boolean replaceExisting : falseAndTrue) {
+                    for (boolean targetExists : falseAndTrue) {
+                        list.add(Arguments.of(type, mode, replaceExisting,
+                                              targetExists));
+                    }
+                }
+            }
+        }
+
+        return list.stream();
+    }
+
+    @ParameterizedTest
+    @MethodSource("params")
+    void copyFollow(PathType type, String mode, boolean replaceExisting,
+                    boolean targetExists) throws IOException {
+        op(OpType.COPY, type, mode, replaceExisting, targetExists, true);
+    }
+
+    @ParameterizedTest
+    @MethodSource("params")
+    void copyNoFollow(PathType type, String mode, boolean replaceExisting,
+                    boolean targetExists) throws IOException {
+        op(OpType.COPY, type, mode, replaceExisting, targetExists, false);
+    }
+
+    @ParameterizedTest
+    @MethodSource("params")
+    void move(PathType type, String mode, boolean replaceExisting,
+              boolean targetExists) throws IOException {
+        op(OpType.MOVE, type, mode, replaceExisting, targetExists, false);
+    }
+
+    void op(OpType op, PathType type, String mode, boolean replaceExisting,
+            boolean targetExists, boolean followLinks) throws IOException {
+
+        Path source = null;
+        Path target = null;
+        Path linkTarget = null;
+        try {
+            switch (type) {
+                case FILE ->
+                    source = Files.createTempFile("file", "dat");
+                case DIR ->
+                    source = Files.createTempDirectory("dir");
+                case LINK -> {
+                    linkTarget = Files.createTempFile("link", "target");
+                    Path link = Path.of("link");
+                    source = Files.createSymbolicLink(link, linkTarget);
+                }
+            }
+
+            if (Files.getFileStore(source).supportsFileAttributeView("posix")) {
```

Comment:
The `@requires` at the top of the test restricts execution to Linux and macOS for now. I would have thought it's a hard requirement that the file system support POSIX file permissions.

Reply:
Moved requirement to top level in cd644e46fed22a1f8b4932b8ee5b54626134ad18. The `@requires` remains as-is.